options/options.go

```diff
@@ -37,3 +41,21 @@ type BuilderOpts struct {
 	// ExpectedBuilderID is the builderID passed in from the user to be verified
 	ExpectedID *string
 }
+
+// VerifierOpts are the options for the verifier, created with the VerifierOptioner functions.
+// In the future, this can include a logger and a rekor client,
+// or be a standalone argument to the verifier functions.
+type VerifierOpts struct {
+	// SigstoreTufClient is the Sigstore TUF client, used for retrieving the Npmjs public keys
+	SigstoreTUFClient apiUtils.SigstoreTUFClient
+}
+
+// VerifierOptioner is a function that sets options for the verifier.
```

Comment:
Is there any reason to use this approach for creating options as opposed to something like:
```
verifierOpts := &options.VerifierOpts{
  SigstoreTUFClient: client,
  ... // Anything else that needs to be set up
}
```
in the client (and then potentially a `DefaultVerifierOptions` func if needed - see comment below)? I think that would be more consistent with `builderOpts`/`provenanceOpts` and clearer for clients to understand?

Reply:
I agree it would be more clear to understand, but my reason is so that we won't me merging in a breaking change at this time.

Reply:
Using variadic seems OK if we need to not make a breaking change to the api. In terms of using an `Optioner` vs something like `options.VerifierOpts`, using the latter might be better to keep consistent with the way we provide the other options (possibly including a `DefaultVerifierOptions` function to make it easy for clients to get the standard set of options). Or are there pros for using an `Optioner` that I'm missing? (I'm not very familiar with that pattern sorry)

Reply:
I'm also new to the pattern, but I see it's used a bit in sigstore-go for making optional arguments. (I prefer to call the returned func an "optioner", instead of an "option".)
 - https://github.com/sigstore/sigstore-go/blob/87bfa5b1c9fcf161c773559f3b470a36ddea4552/pkg/verify/signed_entity.go#L123
 - https://github.com/sigstore/sigstore-go/blob/87bfa5b1c9fcf161c773559f3b470a36ddea4552/cmd/conformance/main.go#L306

I could change it to accept a variadic `options.VerifierOpts`, but then I'd need code to check that the receiver has no more than one of `options.VerifierOpts`. This seems problematic. And If you recall, that was my original approach, but now it seems we could agree the variadic option funcs seem to be a good compromise to making a whole new `VerifyNpmPackageWithSigstoreTUFClient`.

```
func (v *GCBVerifier) VerifyNpmPackage(ctx context.Context,
	attestations []byte, tarballHash string,
	provenanceOpts *options.ProvenanceOpts,
	builderOpts *options.BuilderOpts,
	verifierOpts ...options.VerifierOpts,
) ([]byte, *utils.TrustedBuilderID, error) {
    if len(verifierOpts) > 1 {
        panic("There should be at most one instance of verifierOpts supplied.")
    }
    ...
}
```

Reply:
Hmm yeah I agree that checking there's at most one `options.VerifierOpts` isn't the nicest.

Where I'm getting stuck is that ideally the API function and its arguments would provide a "well-lit path" to users i.e. the function signature would make it easy for clients to:
a) understand what the function is doing
b) call the function correctly/with their desired arguments.

I think right now it's more complex than it needs to be. There's three different sets of options you have to pass in and the options themselves are specified two different ways. I know that's a by-product of hacking on the existing API (which I don't believe was necessarily designed to be part of an external library).

I don't want to block this change from merging - I know it's been in the works for a while now and I definitely think this restructuring to allow a user to pass in a TUF client is a big improvement. Maybe for the future it be good to either:

- Introduce an additional function with a simpler signature (I'm curious to hear Laurent explain why that might be a bad idea)
- Make a breaking change to this API

We should also make sure the example documentation is super clear.

Reply:
@loosebazooka  It might be easier to see this example from an earlier commit
 - https://github.com/slsa-framework/slsa-verifier/pull/768/commits/e4c034adebe6b3e2fb72f7a6e200f38370217105#diff-2b75bff58ca87cf24fae495867d0eb3a039b63b9505b885b705ea55720777cbcR330-R339

Here I was accepting a variadic `...utils.SigstoreTufClient`, which is really a slice of clients. And so it needed an awkward input validation, to make sure the caller was only passing in exactly one client.

```
func (v *GHAVerifier) VerifyNpmPackage(ctx context.Context,
	attestations []byte, tarballHash string,
	provenanceOpts *options.ProvenanceOpts,
	builderOpts *options.BuilderOpts,
	sigstoreTufClient ...utils.SigstoreTufClient,
) ([]byte, *utils.TrustedBuilderID, error) {
	if len(sigstoreTufClient) != 1 {
		return nil, nil, fmt.Errorf("%w: length of variadic sigstoreTufClient must be exactl 1", serrors.ErrorInternal)
	}
    ...
```

And in the previous post `options.VerifierOpts` is a struct containing the actual options. Instead of a variadic `...options.VerifierOpts`, which requires another awkward check, I chose to pass in a variadic `...options.VerifierOptioner`, which is a slice of functions that will set the actual options in a `options.VerifierOpts` structs that is internal to the receiver.

Reply:
oh right, I forgot go doesn't allow function overloading. But are we only using `...` so that we can send 0 or 1 options and not break old users ? It seems like maybe we should be okay breaking the API for this. It'll get caught at compile time and is an easy fix? Get rid of the variadic and force users to provide options or empty.

Reply:
Big +1, I think that would make the API much cleaner.

Reply:
I agree, but we should wait to do a more thorough redesign before breaking the API for this feature.

Reply:
Ah gotcha, okay so if we're being strict about API breakages, I think just using variadic with VerifierOpts kind of leads into the v3 api better, even if we *must* to check for for 0 or 1 at runtime right now (and we feel it's gross)

Using `VerifierOptioner` temporarily and then moving to a new API that accepts VerifierOpts or Empty would be a more significant change?

Reply:
I could do that. @slugclub wdyt?

Reply:
Yep I agree with @loosebazooka that if we need to be strict about API breakages, the best option would be to accept a variadic `VerifierOpts` and check for 0 or 1 at runtime.

Reply:
@slugclub Thanks! I've added it and renamed it to `ClientOpts`, because now we have a separate `VerificationOpts`. Although, I could fold it into `VerificationOpts`, instead. That might make more sense. wdyt @loosebazooka ?

 - https://github.com/slsa-framework/slsa-verifier/pull/768/files#diff-76f121a3a81c8945dc0e0c81a4b95e941d252522eba8704a1aa6f6a1722a8b68R62

Reply:
Yeah sounds good. Thanks everyone for the fruitful discussion here too.

Reply:
Just making sure: you'd prefer it folded into `ClientOpts` folded into `VerificationOpts`?

Reply:
I don't feel super strongly about this, but I think these configurations as separate structs seems reasonable, as we can grow `ClientOpts` to include configuration for other service dependencies of slsa-verifier while `VerificationOpts` is focused on cryptographic verification options.